### PR TITLE
opam generation: add dune dependency if not supplied

### LINKED ISSUE #2289: opam file generation: dune constraint should fill in the version based on dune-project
Noticed this while porting some projects to dune 1.10. With a dune-project like:

```
(lang dune 1.10)

(package
 (name github-unix)
 (depends
  (ocaml (>= 4.03.0))
  dune
  <...>
```

The opam file generated has:

```
depends: [ "dune" ]
```

It should instead have:

```
depends: [ "dune" {>="1.10"} ]
```

based on the `lang 1.10` stanza in dune-project

### PR DESCRIPTION
**Pull request by avsm**
The one guaranteed dependency when generating opam files is dune itself, so add it in with a constraint based on the dune-project language version.

If there is a manually supplied dune dependency specified in the dune-project, then treat that as an override and do not modify it.

closes #2289

### CODE REVIEW

**Comment by avsm** [comment]
A bit of a hack in 6c7660645b884324e43e78b1a23285fe923f7618 -- need to skip this if the project name is... dune :-)

**Review by ghost — APPROVED** [review]

**Comment by diml** [comment]
I wouldn't call it a hack, surely it's ok for Dune to hardcode its own name :)

**Comment by rgrinberg** [comment]
Same question as the other PR: why limit the fix to 1.11?

File: src/opam_create.ml [hunk 1]
```diff
@@ -68,6 +68,28 @@ let package_fields
...
+          Uop (Gte, QVar dune_version) in
+        { Package.Dependency.
+          name = dune_name
+        ; constraint_ = Some constraint_
+        }
+      in
+      let is_dune_depend (pkg : Package.Dependency.t) =
+        Package.Name.equal pkg.name dune_dep.name in
+      if List.exists package.depends ~f:is_dune_depend then
+        package
```
Comment (rgrinberg) [inline src/opam_create.ml]: Does it make sense to check if the constraint is correct?

Comment (avsm) [reply src/opam_create.ml]: I figured that if there is a manual override, then there is a good reason for it (e.g. perhaps to exclude a particular dune version). Checking those isn't easy, so it's best to leave a human-written one as-is.

**Comment by avsm** [comment]
With promoted files, the intention is to not change behaviour until the user bumps the dune version. I followed @diml's lead on that with other features in 1.11 (like the new `build:` stanza that only activates at 1.11)

**Comment by rgrinberg** [comment]
Okay, it's merged.

I guess I'm interpreting this more of a bugfix than a new feature and it seems a bit cruel not to provide the fix for everyone.

**Comment by avsm** [comment]
I'm open to changing that if you think it better. But I worry that it'll result in "promote flaps" if users switch between dune versions for whatever reason.  It does need a docs note as I agree that it might be confusing if users don't know to use 1.11 instead of 1.10 in their dune project file.

**Comment by rgrinberg** [comment]
My main worry is as follows:

* We don't actually have a consistent policy of preserving the old behavior when we introduce something that is a strict improvement. Sometimes it's too time consuming to maintain all the different versions.

* Once we generate opam files on the fly, it seems that it's best to generate the most "correct" version of this file.

I think that your proposed behavior is more correct with promoted files. Let's keep it for now.

**Comment by diml** [comment]
The official policy is: we should never break the build of existing released packages when changing Dune, except for packages that use experimental features. When preserving full backward compatibility is too difficult for some reason, we should discuss between us to see what can be done to move forward and mitigate the breakages.

In this case, changing the formatting wouldn't break existing released packages. However, having the contents of source files tied to the version of a tool installed on the system doesn't feel right. If it was too difficult to version the contents of generated opam files, then we should indeed not do it. Looking at the current code, the level of complexity seems acceptable to me.

I do realise that we are setting pretty high standards for an open source project, and many other projects don't go that far regarding backward complexity. I'm only asking this because I know that we have an excellent development team formed of programmers who can definitely handle this level of complexity.

> Once we generate opam files on the fly, it seems that it's best to generate the most "correct" version of this file.

Agreed. For opam files that will be generated on the fly, we will indeed be free to generate the most correct version regardless of the version of the dune language used by the project.

**Pull request closed without merging**